modules/writing_observer/writing_observer/awe_nlp.py

```diff
@@ -202,52 +203,99 @@ async def process_texts_parallel(texts, options=None):
-        annotated_texts = await processor[mode](just_the_text, list(needed_options))
-
-        for annotated_text, single_doc in zip(annotated_texts, need_processing):
-            if annotated_text != "Error":
-                single_doc.update(annotated_text)
-                text_hash = 'NLP_CACHE_' + learning_observer.util.secure_hash(single_doc['text'].encode('utf-8'))
-                new_cache = {k: v for k, v in single_doc.items() if k != 'student'}
-                await cache.set(text_hash, new_cache)
-        results.extend(need_processing)
+
+        features_running = set(json.loads(text_cache_data['features_running'])) if 'features_running' in text_cache_data else set()
+        text_cache_data.setdefault('features_available', dict())
+
+        # Check if some options are a subset of features_available
+        if text_cache_data['features_available']:
+            found_options = set_options.intersection(text_cache_data['features_available'].keys())
+            writing.update(text_cache_data['features_available'])
+
+            # If all options were found
+            if found_options == set_options:
+                results.append(writing)
+                continue
+
+        # Check if some options are a subset of features_running
+        # features that are needed but are already running
+        run_successful = False
+        not_found = set_options - found_options
+        features_needed_running = set()  # Features that are needed but are already processing
+        if features_running:
+            features_needed_running = not_found.intersection(features_running)
+        if len(features_needed_running) > 0:
+            while True:
+                new_cache = await cache[text_hash]
+                if new_cache['stop_time'] != "running":
+                    run_successful = True
+                    break
+                if (datetime.datetime.now() - datetime.datetime.strptime(new_cache['start_time'], "%Y-%m-%d %H:%M:%S.%f")).total_seconds() > running_features_wait_time:
+                    break
+                await asyncio.sleep(recurring_sleep_time)
+
+            if run_successful:
+                # features_running will be available in features_available after they finish running.
+                writing.update(text_cache_data['features_available'])
+                found_options = found_options.union(features_needed_running)
+                features_needed_running = set()
+                # If all options are found
+                if found_options == set_options:
+                    results.append(writing)
+                    continue
+
+        # Add not found options to features_running and update cache
+        not_found = set_options - found_options
+        features_running = not_found
+        text_cache_data['features_running'] = json.dumps(list(features_running))
+        text_cache_data['start_time'] = str(datetime.datetime.now())
```

Comment:
We shouldn't reinvent how we do timestamps throughout the code. We've standardized on `datetime.datetime.utcnow().isoformat()`. The best steps here would be to move this code from the merkle_tree.py file to util.py file, and use the helper here. 

Conversely, parsing should probably be done with a helper which calls `dateutil.parser.isoparse`.

Reply:
Thanks for your review Piotr. I will make the changes you suggested and push out a new commit.